src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -379,6 +367,24 @@ internal void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Bu
 
             lock (_sync)
             {
+                // If the reader is completed
+                if (_readerCompletion.IsCompleted)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+                }
+
+                var fullBuffer = new ReadOnlySequence<byte>(_readHead, _readHeadIndex, _commitHead, _commitHeadIndex - _commitHead.Start);
+                var examinedToConsumed = fullBuffer.Slice(examined, consumed);
+                if (!SequenceMarshal.TryGetReadOnlySequenceSegment(
+                    examinedToConsumed,
+                    out ReadOnlySequenceSegment<byte> examinedSegment,
+                    out int examinedIndex,
+                    out ReadOnlySequenceSegment<byte> consumedSegment,
+                    out int consumedIndex))
+                {
+                    ThrowHelper.ThrowInvalidOperationException_AlreadyReading();
```

Comment:
Why does this mean you're already reading?

Reply:
Temp, should have put WIP on this PR

Reply:
Decided to fix this last minute.